# Imports

In [1]:
import pandas as pd

from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

from collections import Counter

# Dataset Inspection

In [2]:
# Load the cleaned data
df = pd.read_csv('clean_df.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2892 entries, 0 to 2891
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   quantity         2586 non-null   float64
 1   unit             2524 non-null   object 
 2   ingredient       2890 non-null   object 
 3   ingredient_step  2892 non-null   int64  
 4   recipe_id        2892 non-null   int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 113.1+ KB


In [4]:
# Group by recipe_id and aggregate ingredients into lists
grouped_df = df.groupby('recipe_id')['ingredient'].apply(list).reset_index()

# Display grouped DataFrame
grouped_df

recipe_id                                         ingredient
0            0  [dark rum, lime juice, sugar, angostura, perno...
1            1  [silver tequila, marie brizard creme de cacao,...
2            2  [gin, mari brizard white creme de cacao, lille...
3            3                             [pernod, sugar, water]
4            4  [gold tequila, gold rum, grapefruit juice, pin...
..         ...                                                ...
650        650      [light rum, galliano, triple sec, lime juice]
651        651  [dark rum, light rum, tia maria, orange juice,...
652        652     [vodka, galliano, lime juice, pineapple juice]
653        653        [yellow chartreuse, pernod, apricot brandy]
654        654  [rye whiskey, lemon juice, pineapple syrup, do...

[655 rows x 2 columns]

In [5]:
# Extract the list of lists of ingredients
recipes = grouped_df['ingredient'].tolist()

# Get the list of all unique ingredients
all_ingredients = df['ingredient'].unique().tolist()

# Use MultiLabelBinarizer to encode the ingredients
mlb = MultiLabelBinarizer(classes=all_ingredients)
encoded_recipes = mlb.fit_transform(recipes)

# Create a DataFrame for easier manipulation
encoded_df = pd.DataFrame(encoded_recipes, columns=mlb.classes_)

In [6]:
encoded_df.shape

(655, 389)

# Outline:

Model Building:

Our model architecture is:

l0: 389 (amount of unique ingredients)

hidden layers:          relu

L :            sigmoid   , 389

In [7]:
input_dim = len(all_ingredients)  # Number of unique ingredients

In [ ]:


model = Sequential()
model.add(Dense(units=128, input_dim=input_dim, activation='relu'))
model.add(Dense(units=64, activation='relu'))
model.add(Dense(units=input_dim, activation='sigmoid'))  # Output layer with sigmoid activation

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [ ]:
# attempt at at a 2nd model, larger architecture since 389 param

input_dim = len(all_ingredients)  # Number of unique ingredients

model = Sequential()
model.add(Dense(units=256, input_dim=input_dim, activation='relu'))  # First hidden layer
model.add(Dense(units=128, activation='relu'))  # Second hidden layer
model.add(Dense(units=64, activation='relu'))  # Second hidden layer
model.add(Dense(units=128, activation='relu'))  # Second hidden layer
model.add(Dense(units=256, activation='relu'))  # Third hidden layer
model.add(Dense(units=input_dim, activation='sigmoid'))  # Output layer

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [ ]:
# Print model summary to verify the architecture
model.summary()

In [8]:
# Attemping dropout layers
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

# Define the model
model = Sequential()
model.add(Dropout(0.1, input_shape=(input_dim,)))  # Dropout layer for the input layer
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.5))  # Dropout layer for hidden layers
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))  # Dropout layer for hidden layers
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))  # Dropout layer for hidden layers
model.add(Dense(input_dim, activation='sigmoid'))  # Output layer for multi-label classification


# Compile the model
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dropout (Dropout)           (None, 389)               0         
                                                                 
 dense (Dense)               (None, 512)               199680    
                                                                 
 dropout_1 (Dropout)         (None, 512)               0         
                                                                 
 dense_1 (Dense)             (None, 256)               131328    
                                                                 
 dropout_2 (Dropout)         (None, 256)               0         
                                                                 
 dense_2 (Dense)             (None, 128)               32896     
                                                                 
 dropout_3 (Dropout)         (None, 128)               0

Model Training:

In [9]:
X = encoded_df.values  # Input features (binary vectors of ingredients)
y = encoded_df.values  # Target labels (same as input for multi-label classification)

#model.fit(X, y, epochs=20, batch_size=50, validation_split=0.2)

Generating New Recipes:

Start with a seed ingredient or set of ingredients.
Use the trained model to predict additional ingredients.

In [ ]:
model.fit(X, y, epochs=50, batch_size=50, validation_split=0.2)

seed_ingredients = ['vodka']  # Example seed ingredients
seed_vector = mlb.transform([seed_ingredients])[0]

predicted_probabilities = model.predict(seed_vector.reshape(1, -1))
predicted_ingredients = mlb.inverse_transform((predicted_probabilities > 0.00001).astype(int))[0]

generated_recipe = list(set(seed_ingredients + list(predicted_ingredients)))
print("Generated Recipe:", generated_recipe)


In [10]:
# Number of times to generate recipes
num_iterations = 100

# Example seed ingredients
seed_ingredients = ['gin', 'lime juice']
seed_vector = mlb.transform([seed_ingredients])[0]

# List to store generated recipes
generated_recipes = []

for _ in range(num_iterations):
    # Fit the model for one epoch
    model.fit(X, y, epochs=5, batch_size=389, validation_split=0.2, verbose=0)
    
    # Predict probabilities
    predicted_probabilities = model.predict(seed_vector.reshape(1, -1))
    
    # Convert probabilities to ingredients
    predicted_ingredients = mlb.inverse_transform((predicted_probabilities > 0.001).astype(int))[0]
    
    # Generate recipe
    generated_recipe = list(set(seed_ingredients + list(predicted_ingredients)))
    
    # Store the generated recipe as a tuple for easier counting
    generated_recipes.append(tuple(generated_recipe))

# Count the occurrences of each unique recipe
recipe_counts = Counter(generated_recipes)

# Convert to DataFrame
df = pd.DataFrame(recipe_counts.items(), columns=['Recipe', 'Count'])

# Sort by count (optional)
df = df.sort_values(by='Count', ascending=False)

print(df)


1/1 [==============================] - 0s 16ms/step
                                               Recipe  Count
0   (nan, pineapple pieces, freshly cracked pepper...      6
84  (dark rum, mandarine napoléon, cranberry juice...      2
59  (st. germain, simple syrup, green chatreuse, o...      1
68  (st. germain, simple syrup, black sambuca, gre...      1
67  (st. germain, simple syrup, green chatreuse, o...      1
..                                                ...    ...
29  (nan, pineapple pieces, vermouth, st. germain,...      1
28  (nan, pineapple pieces, vermouth, st. germain,...      1
27  (nan, pineapple pieces, vermouth, st. germain,...      1
26  (nan, pineapple pieces, vermouth, st. germain,...      1
93  (dark rum, mandarine napoléon, cranberry juice...      1

[94 rows x 2 columns]


In [11]:
df['Recipe']

0     (nan, pineapple pieces, freshly cracked pepper...
84    (dark rum, mandarine napoléon, cranberry juice...
59    (st. germain, simple syrup, green chatreuse, o...
68    (st. germain, simple syrup, black sambuca, gre...
67    (st. germain, simple syrup, green chatreuse, o...
                            ...                        
29    (nan, pineapple pieces, vermouth, st. germain,...
28    (nan, pineapple pieces, vermouth, st. germain,...
27    (nan, pineapple pieces, vermouth, st. germain,...
26    (nan, pineapple pieces, vermouth, st. germain,...
93    (dark rum, mandarine napoléon, cranberry juice...
Name: Recipe, Length: 94, dtype: object

In [ ]:
#df['Recipe'][0]

In [12]:
import numpy as np

# Assuming `predicted_probabilities` is already defined and is a 2D array
# and `mlb` is your MultiLabelBinarizer instance

# Get the indices that would sort each row of the array in descending order
sorted_indices = np.argsort(-predicted_probabilities, axis=1)

# Get the sorted probabilities
sorted_probabilities = np.take_along_axis(predicted_probabilities, sorted_indices, axis=1)

# For each row, map the sorted indices back to the ingredient names
# Assuming you want to do this for the first row in the predictions
top_ingredients_indices = sorted_indices[0]
top_probabilities = sorted_probabilities[0]

# Map these indices back to the ingredient names
top_ingredients = mlb.classes_[top_ingredients_indices]

# Print the top ingredients and their corresponding probabilities
for ingredient, probability in zip(top_ingredients, top_probabilities):
    print(f"Ingredient: {ingredient}, Probability: {probability}")


Ingredient: gin, Probability: 0.9867560267448425
Ingredient: lime juice, Probability: 0.9344647526741028
Ingredient: grenadine, Probability: 0.03838746249675751
Ingredient: apricot brandy, Probability: 0.03816235810518265
Ingredient: luxardo maraschino, Probability: 0.030050097033381462
Ingredient: cointreau, Probability: 0.025867082178592682
Ingredient: lemon juice, Probability: 0.021835507825016975
Ingredient: simple syrup, Probability: 0.0213521346449852
Ingredient: green chatreuse, Probability: 0.020899588242173195
Ingredient: orange curaçao, Probability: 0.017704429104924202
Ingredient: orange juice, Probability: 0.014828570187091827
Ingredient: clam juice, Probability: 0.013319894671440125
Ingredient: rum, Probability: 0.009371038526296616
Ingredient: red pepper sauce, Probability: 0.009366866201162338
Ingredient: green chartreuse, Probability: 0.008959535509347916
Ingredient: grapefruit juice, Probability: 0.008716586977243423
Ingredient: guava nectar, Probability: 0.00819884147

Summary of Steps:
Clean Data Preparation:

Create a DataFrame with binary encoded ingredient vectors for each recipe.
Model Building:

Define a neural network to learn from these vectors.
Model Training:

Train the model on your dataset.
Recipe Generation:

Generate new recipes starting from seed ingredients.
By following these steps, you should be able to create and train a model that can generate new cocktail recipes based on the patterns it learns from your dataset. Let me know if you need more detail on any of these steps!

The architecture described here is a simple feedforward neural network, also known as a Multi-Layer Perceptron (MLP). This architecture is used for multi-label classification, where each ingredient is treated as a binary label that can be present or absent in a recipe.

Architecture Explanation
Input Layer:

The input layer consists of nodes equal to the number of unique ingredients. Each node represents whether a particular ingredient is present (1) or absent (0) in the recipe.
Input Dimension: input_dim = len(all_ingredients).
Hidden Layers:

The network has two hidden layers, each with 128 neurons. These layers use the ReLU (Rectified Linear Unit) activation function, which introduces non-linearity to the model and allows it to learn complex patterns.
First Hidden Layer: Dense(units=128, input_dim=input_dim, activation='relu')
Second Hidden Layer: Dense(units=128, activation='relu')
Output Layer:

The output layer also consists of nodes equal to the number of unique ingredients. Each node represents the probability of the corresponding ingredient being part of the recipe.
The sigmoid activation function is used in the output layer to produce probabilities between 0 and 1 for each ingredient.
Output Layer: Dense(units=input_dim, activation='sigmoid')
Loss Function:

The model uses binary cross-entropy loss, suitable for multi-label classification where each label (ingredient) is a binary decision (present or absent).
Loss Function: binary_crossentropy
Optimizer:

The Adam optimizer is used to minimize the loss function during training. Adam is a popular choice due to its adaptive learning rate and efficiency.
Optimizer: adam
Metrics:

The accuracy metric is used to evaluate the model's performance during training and validation.
Metrics: accuracy
Why This Architecture?
Simplicity: This architecture is straightforward and easy to implement, making it suitable for a dataset of your size (654 recipes and 447 unique ingredients).
Flexibility: The use of dense layers allows the network to learn from the presence and absence of each ingredient, capturing the relationships between different ingredients.
Scalability: Adding more hidden layers or increasing the number of neurons in each layer can improve the model's capacity to learn more complex patterns, if needed.
Summary
This architecture is a feedforward neural network designed for multi-label classification. It takes a binary vector representing the presence or absence of each ingredient as input and outputs a binary vector representing the predicted probabilities of each ingredient being part of the cocktail recipe. This approach is suitable for generating new recipes by learning the common ingredient combinations from the training data.








## Next chunk of code: Randomly decides to include this ingredient based on its probability

In [18]:
import numpy as np
from collections import Counter
import pandas as pd

# Assuming `predicted_probabilities` is already defined and is a 2D array
# and `mlb` is your MultiLabelBinarizer instance

def generate_semi_random_cocktail(predicted_probabilities, mlb, top_n=10, threshold=0.05):
    # Ensure predicted_probabilities is 2D
    if predicted_probabilities.ndim == 1:
        predicted_probabilities = predicted_probabilities.reshape(1, -1)
    
    # Get the indices that would sort each row of the array in descending order
    sorted_indices = np.argsort(-predicted_probabilities, axis=1)
    
    # Get the sorted probabilities
    sorted_probabilities = np.take_along_axis(predicted_probabilities, sorted_indices, axis=1)

    # For each row, map the sorted indices back to the ingredient names
    # Assuming you want to do this for the first row in the predictions
    top_ingredients_indices = sorted_indices[0]
    top_probabilities = sorted_probabilities[0]

    # Map these indices back to the ingredient names
    top_ingredients = mlb.classes_[top_ingredients_indices]

    # Semi-random selection based on sorted probabilities
    selected_ingredients = []
    for i in range(top_n):
        if top_probabilities[i] > threshold:
            # Randomly decide to include this ingredient based on its probability
            if (np.random.random() * 0.1) < top_probabilities[i]:
                selected_ingredients.append(top_ingredients[i])
    
    return selected_ingredients

# Number of times to generate recipes
num_iterations = 100

# Example seed ingredients
seed_ingredients = ['gin', 'lime juice']
seed_vector = mlb.transform([seed_ingredients])[0]

# List to store generated recipes
generated_recipes = []

for _ in range(num_iterations):
    # Fit the model for one epoch
    model.fit(X, y, epochs=1, batch_size=389, validation_split=0.2, verbose=0)
    
    # Predict probabilities
    predicted_probabilities = model.predict(seed_vector.reshape(1, -1))
    
    # Generate recipe using the semi-random selection method
    generated_recipe = generate_semi_random_cocktail(predicted_probabilities, mlb, top_n=10, threshold=0.000000005)
    
    # Add the seed ingredients to ensure they are included
    full_recipe = list(set(seed_ingredients + generated_recipe))
    
    # Store the generated recipe as a tuple for easier counting
    generated_recipes.append(tuple(full_recipe))

# Count the occurrences of each unique recipe
recipe_counts = Counter(generated_recipes)

# Convert to DataFrame
df = pd.DataFrame(recipe_counts.items(), columns=['Recipe', 'Count'])

# Sort by count (optional)
df = df.sort_values(by='Count', ascending=False)

print(df)


1/1 [==============================] - 0s 14ms/step
                                               Recipe  Count
7                                   (lime juice, gin)     21
4               (lime juice, luxardo maraschino, gin)     10
22                 (lime juice, gin, green chatreuse)      8
10                  (apricot brandy, lime juice, gin)      8
5   (apricot brandy, lime juice, gin, green chatre...      5
24                    (lime juice, simple syrup, gin)      4
17    (apricot brandy, lime juice, simple syrup, gin)      3
21                       (cointreau, lime juice, gin)      3
25   (cointreau, lime juice, luxardo maraschino, gin)      3
14                          (lime juice, simple, gin)      3
13      (lime juice, simple, luxardo maraschino, gin)      3
18  (gin, green chatreuse, lime juice, simple, apr...      2
23       (apricot brandy, cointreau, lime juice, gin)      2
11  (apricot brandy, lime juice, luxardo maraschin...      2
9    (lime juice, luxardo marasch

In [19]:
df

Recipe  Count
7                                   (lime juice, gin)     21
4               (lime juice, luxardo maraschino, gin)     10
22                 (lime juice, gin, green chatreuse)      8
10                  (apricot brandy, lime juice, gin)      8
5   (apricot brandy, lime juice, gin, green chatre...      5
24                    (lime juice, simple syrup, gin)      4
17    (apricot brandy, lime juice, simple syrup, gin)      3
21                       (cointreau, lime juice, gin)      3
25   (cointreau, lime juice, luxardo maraschino, gin)      3
14                          (lime juice, simple, gin)      3
13      (lime juice, simple, luxardo maraschino, gin)      3
18  (gin, green chatreuse, lime juice, simple, apr...      2
23       (apricot brandy, cointreau, lime juice, gin)      2
11  (apricot brandy, lime juice, luxardo maraschin...      2
9    (lime juice, luxardo maraschino, gin, grenadine)      2
6                        (lime juice, gin, grenadine)      2
2       (grenadine, lime juice, gin, green chatreuse)      2
29               (lime juice, simple, gin, grenadine)      1
28      (lime juice, green chatreuse, gin, grenadine)      1
27      (cointreau, lime juice, gin, green chatreuse)      1
31         (lime juice, simple, gin, green chatreuse)      1
26  (lime juice, simple syrup, luxardo maraschino,...      1
32                    (lime juice, gin, orange juice)      1
30  (cointreau, simple syrup, gin, green chatreuse...      1
0   (gin, green chatreuse, lime juice, apricot bra...      1
20  (lime juice, green chartreuse, gin, green chat...      1
19  (cointreau, gin, green chatreuse, lime juice, ...      1
1   (lime juice, orange curaçao, luxardo maraschin...      1
16  (lime juice, orange curaçao, gin, green chatre...      1
15  (gin, lime juice, simple, orange curaçao, apri...      1
12          (apricot brandy, lime juice, simple, gin)      1
8                   (lime juice, orange curaçao, gin)      1
3   (gin, green chatreuse, lime juice, orange cura...      1
33                     (lime juice, lemon juice, gin)      1

In [ ]:
all_ingredients[1]

In [ ]:
print("Predicted Probabilities:", predicted_probabilities)

In [ ]:
predicted_probabilities[86]